crates/red_knot_python_semantic/src/types.rs

```diff
@@ -4883,6 +4885,79 @@ impl<'db> Type<'db> {
             | Self::AlwaysFalsy => None,
         }
     }
+
+    /// Returns a tuple of two spans. The first is
+    /// the span for the identifier of the function
+    /// definition for `self`. The second is
+    /// the span for the return type in the function
+    /// definition for `self`.
+    ///
+    /// If there are no meaningful spans, then this
+    /// returns `None`. For example, when this type
+    /// isn't callable or if the function has no
+    /// declared return type.
+    fn return_type_span(&self, db: &'db dyn Db) -> Option<(Span, Span)> {
```

Comment:
This API and `parameter_span` can both cause cross-module dependencies. We already discussed this with the improved invalid-call diagnostic, and I don't think there's anything to be done about that core fact -- it's just a reality of providing nice diagnostics.

I'm not sure if it would be useful to call this out in the doc comment, or via some kind of naming convention? I don't really think it's likely we'd attempt to use these methods unless we really need them for diagnostics, so tbh I'm not sure how much that would benefit us.

Reply:
I added this as a doc comment to this method and `parameter_span`:

```
    /// # Performance
    ///
    /// Note that this may introduce cross-module
    /// dependencies. This can have an impact on
    /// the effectiveness of incremental caching
    /// and should therefore be used judiciously.
    ///
    /// An example of a good use case is to improve
    /// a diagnostic.
```

I think the cross module dependency thing is really subtle, so a naming convention might actually help here. But yeah, I also don't have a good feel for how much of a problem this will be in practice.